Add "Stream details" section in stream settings

tnmkr opened the pull request:
CZO discussions:
[#feedback](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Creator.20of.20Stream.20info/near/1571036), [#design](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Stream.20details.22.20section.20design), [#backend](https://chat.zulip.org/#narrow/stream/3-backend/topic/Type.20for.20Stream.2Ecreator.20field)

Fixes: #25648 

- [x] Add a "Stream details" section to the bottom of the General tab in stream settings.
- [x] Put the email address in that section (i.e., "Email address" will be a field name, and not a top-level section heading)
- [x] Add "Stream ID" and its value (can be a follow-up)
- [x] Adjust the tips in https://github.com/zulip/zulip/pull/27941 to reflect these changes.
- [x] Add the following line.
> Created by @user on date. [Stream events](https://github.com/zulip/zulip/issues/25648)

<details><summary>git grep on css classes</summary>
<p>

```html
(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream-email-box
web/src/stream_ui_updates.js:        $settings.find(".stream-email-box").show();
web/src/stream_ui_updates.js:        $settings.find(".stream-email-box").hide();
web/templates/stream_settings/stream_settings.hbs:                <div class="input-group stream-email-box">

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream-email-box-header

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream_details_box
web/styles/subscriptions.css:    .stream_details_box {
web/templates/stream_settings/stream_settings.hbs:            <div class="stream_details_box">
web/templates/stream_settings/stream_settings.hbs:                <div class="stream_details_box_header">

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream_details_box_header
web/templates/stream_settings/stream_settings.hbs:                <div class="stream_details_box_header">

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep copy_email_button
web/src/stream_edit.js:    $("#streams_overlay_container").on("click", ".copy_email_button", (e) => {
web/templates/stream_settings/stream_settings.hbs:                    <button class="button small rounded copy_email_button" id="copy_stream_email_button" type="button">

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream_creator_details
web/styles/input_pill.css:.stream_creator_details > .pill-container {
web/styles/input_pill.css:.stream_creator_details > .pill-container {
web/templates/stream_settings/stream_settings.hbs:                <div class="stream_creator_details stream_details_subsection">

(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep stream_details_subsection
web/styles/subscriptions.css:        > .stream_details_subsection {
web/templates/stream_settings/stream_settings.hbs:                <div class="stream_creator_details stream_details_subsection">
web/templates/stream_settings/stream_settings.hbs:                <div class="stream_details_subsection">
```

</p>
</details> 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:** 
| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/43791a1e-77db-4279-9f6b-6e7bf45ce1e7) | ![image](https://github.com/zulip/zulip/assets/133781250/615fe912-1efe-4197-8dfc-26f774dfd140) |
| Light mode | Dark mode |
| ![image](https://github.com/zulip/zulip/assets/133781250/4f675b2f-4ef6-4fd7-9570-459a0bef0bc2) | ![image](https://github.com/zulip/zulip/assets/133781250/7736f89f-3ebd-4b27-a7c6-2b6fc65e6764) |
| Stream where creator is `null` | "Generate email" error |
| ![image](https://github.com/zulip/zulip/assets/133781250/533f1108-7259-41e0-b670-8cbfc524024e) | ![image](https://github.com/zulip/zulip/assets/133781250/428c3073-2aee-46aa-b74e-138481e54860) |
| Stream where creator is inaccessible, eg for guests | -- |
| ![image](https://github.com/zulip/zulip/assets/133781250/a2d650fd-09d0-4d09-91bc-8a0643de9e65) | -- |
| Docs, before (constructing-narrow, incoming-webhooks) | Docs, after |
| ![image](https://github.com/zulip/zulip/assets/133781250/8f968de0-76a1-408c-a2b4-d6b28769764a) | ![image](https://github.com/zulip/zulip/assets/133781250/16f91619-c3b4-4dd5-8ef0-ad5810e02c33) |
| ![image](https://github.com/zulip/zulip/assets/133781250/ac4ff5dc-6361-4e21-9fd2-36c03e9a58b0) | ![image](https://github.com/zulip/zulip/assets/133781250/3d806059-c14a-4dbb-8a40-1f8d955dfe85) |

Where `user_display_only_pill.hbs` is used.
| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/787fbdee-25fe-4967-a788-4fd9bb3abd14) | ![image](https://github.com/zulip/zulip/assets/133781250/921b8a7f-599a-4178-9552-5d31af9c3642) |
| ![image](https://github.com/zulip/zulip/assets/133781250/0ad6c4f9-be45-4812-9e93-748420e87296) | ![image](https://github.com/zulip/zulip/assets/133781250/93af09ee-c14e-4105-9031-21fc18a2bddc) | 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

tnmkr commented on web/styles/subscriptions.css:
```diff
@@ -950,15 +950,15 @@ h4.user_group_setting_subsection_title {
             vertical-align: top;
             margin-left: 15px;
         }
-
-        .copy_email_button {
-            padding: 10px 15px;
-        }
```
To make the button smaller and better match existing buttons.

timabbott replied:
The `copy_email_button` change would be nice to do as its own prep commit too, since it's not logically part of the creator change.

tnmkr replied:
I have moved this change into its own commit.

tnmkr commented on web/styles/input_pill.css:
```diff
@@ -176,13 +177,26 @@
         }
 
-        > .pill-value {
-            padding: 5px;
+        > .pill-label {
             max-width: 165px;
         }
```
I have confirmed that this doesn't affect visual appearance of these pills.

timabbott replied:
Please do removing that `5px` padding in a prep commit with a proper commit message explaining your motivation for changing it and how you verified it has no effect. `git blame` or similar tools might also be helpful to see why it was added.

tnmkr replied:
I was wrong about this. Removing the `5px` padding did actually remove the spacing b/w the user name and `"(you)"` flair text. To fix that, I have added this padding back. 

I have also moved this into its own prep commit, along with HTML changes to the handlebar parital.

This was added as part of PR that introduced `user_display_only_pill` partial. PR #28488. I am not sure, but looks like the intention was to have spacing similar to what `.pill-label` provides.

timabbott replied:
That sounds right -- maybe worth starting a #frontend thread with @sujalshah-bit to discuss further cleanups to this code path.

tnmkr commented:
I have implemented changes discussed in [CZO #design](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Stream.20details.22.20section.20design), and updated screenshots in PR.

@alya This is ready for review.

alya commented:
Thanks! Looks promising, but I posted a [question](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Stream.20details.22.20section.20design/near/1777770).

alya commented:
In the meantime, @sahil839, perhaps you could start reviewing this one? I haven't tested.

tnmkr commented:
Made the following update as per discussion in [CZO #design](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Stream.20details.22.20section.20design/near/1777770):
- Only show date when there is no creator.

timabbott commented on zerver/models/streams.py:
```diff
@@ -180,6 +181,7 @@ def is_history_public_to_subscribers(self) -> bool:
     # * "realm" and "recipient" are not exposed to clients via the API.
     API_FIELDS = [
         "date_created",
+        "creator_id",
```
Put this field first; the values are clearly intended to be alphabetized.

tnmkr replied:
I have fixed this here and in other places as needed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -19070,6 +19090,19 @@ components:
+          description: |
+            The ID of the user who created this stream.
+
+            Is `null` for streams with no creator. Streams that existed
```
"no recorded creator" would be clearer.

tnmkr replied:
Changed the wording as suggested.

timabbott commented on web/src/stream_data.ts:
```diff
@@ -186,6 +187,14 @@ export function get_sub_by_id(stream_id: number): StreamSubscription | undefined
+        return undefined;
+    }
+
+    return people.get_by_user_id(creator_id);
```
I think this is an appropriate application for `get_user_by_id_assert_valid` -- we know the user exists, we just may not have access to it.

tnmkr replied:
Update to use `get_user_by_id_assert_valid` function.

timabbott commented on web/styles/input_pill.css:
```diff
@@ -162,7 +162,8 @@
 
-.panel_subscriber_member_list > .pill-container {
+.panel_subscriber_member_list > .pill-container,
+.stream_creator_details > .pill-container {
```
It's probably worth doing a prep commit that adds an explicit class name describing what's special about the pills that get this block, and apply that class, rather than just having a list of specific widget class names.

tnmkr replied:
Added a new css class `.display_only_pill` for these pills. 

<details><summary>git grep on new css class</summary>
<p>

```css
(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep ".display_only_pill"
web/styles/input_pill.css:.panel_subscriber_member_list > .display_only_pill,
web/styles/input_pill.css:.stream_creator_details > .display_only_pill {
web/templates/stream_settings/new_stream_user.hbs:        {{> ../user_display_only_pill display_value=full_name}}
web/templates/stream_settings/stream_member_list_entry.hbs:        {{> ../user_display_only_pill display_value=name}}
web/templates/stream_settings/stream_settings.hbs:                            {{> ../user_display_only_pill
web/templates/user_display_only_pill.hbs:<span class="pill-container display_only_pill {{#if is_inline}}pill-container-btn inline_with_text_pill{{/if}}">
```

</p>
</details>

timabbott replied:
This likely deserves further cleanup but we can do that in a follow-up PR.

timabbott commented on web/styles/input_pill.css:
```diff
@@ -178,13 +179,26 @@
 
+/* Specific styling, on top of shared styling above,
+to vertically align the pill with surrounding text. */
+.stream_creator_details > .pill-container {
```
This could maybe benefit from an `inline-with-text-pill` more declarative class, perhaps?

tnmkr replied:
Added new css class `.inline_with_text_pill` for this. We seem to be preferring `_` instead of `-` in new class names, right?

<details><summary>git grep on new class</summary>
<p>

```css
(zulip-py3-venv) vagrant@fbc57b970081:/srv/zulip$ git grep ".inline_with_text_pill"
web/styles/input_pill.css:.stream_creator_details > .pill-container-btn.inline_with_text_pill {
web/templates/user_display_only_pill.hbs:<span class="pill-container display_only_pill {{#if is_inline}}pill-container-btn inline_with_text_pill{{/if}}">
```
</p>
</details>

timabbott replied:
https://zulip.readthedocs.io/en/stable/subsystems/html-css.html#use-clear-unique-names-for-classes-and-object-ids covers this -- we're very slowly moving towards `-`s, but `_`s used to be the default. Not particularly important one way or the other, but FYI.

timabbott commented:
We should do a backfill migration to populate the `creator_id` field using the data that we store in the `RealmAuditLog` table with `STREAM_CREATED = 601`. This would reduce cases where we don't have data greatly. Are you thinking that we should do that as a follow-up PR? https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html and reading recent migrations to see techniques for avoiding inefficient loops is probably worth doing for that project.

If that feels challenging, I think it would be a valid checkpoint to leave that data backfill as a follow-up PR, since this design change does handle the case where we lack data gracefully. But we should avoid claiming in the API documentation that we will never have data for older streams.

My only other larger concern is that I don't like how medium-length user pills get abbreviated; we might want to find a home for "View stream events" on a different line.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 9.0
 
+**Feature level 248**
```
You would also need to bump the `API_FEATURE_LEVEL` field in `version.py`.

tnmkr replied:
Updated in `version.py` and made necessary changes api changelog and docs.

tnmkr commented on web/templates/user_display_only_pill.hbs:
```diff
@@ -1,14 +1,16 @@
+            {{~#if has_status~}}
+            {{~> status_emoji status_emoji_info~}}
+            {{~/if~}}
+        </span>
```
Moved this into a seperate commit.

tnmkr commented:
Update:
- Addressed all review comments.
- Added `[`GET /users/me/subscriptions`](/api/get-subscriptions)` path in `changelog.md`. I had missed new field on `Subscription` object example in `zulip.yaml`, fixed that aswell.
- Updated migration to backfill data from RealmAuditLog
---

> Are you thinking that we should do that as a follow-up PR?

I have added backfilling to the migration in this PR. I think follow-up PRs would also be needed to update imports? I am not sure.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -4413,6 +4437,35 @@ def test_user_settings_for_creating_web_public_streams(self) -> None:
+                continue
+
+            # Streams that aren't created using the api should have no creator
+            self.assertIsNone(stream["creator_id"])
```
It might be worth a follow-up change to `populate_db` to set creators for the initial streams, which might then require changing this test.

tnmkr replied:
Opened #30082 for this. Note that this doesn't fix the issue with `import_realm` tests not failing (mentioned below). 

#30081 added some tests and has some comments regarding that.

timabbott commented on web/styles/input_pill.css:
```diff
@@ -191,6 +192,32 @@ meant to show profile cards when clicked */
+            }
+        }
+    }
+}
```
I'm not sure what the cause is, but the pills sometimes show blue link styling on hover -- can you debug that?

tnmkr replied:
I have reproduced this in dark mode. Probably the best solution here is to use `.display_only_pill` selector instead of `.user_panel_list > .pill_container` in `dark_theme.css`.

I'll resolve this in a cleanup #frontend thread with @sujalshah-bit as suggested above in this PR.

timabbott commented:
This looks pretty great! I pushed the following changes:
- Small wording tweaks to the API documentation.
- Fixed the migration to not try to modify every stream on the server in a single transaction, and also to use the existing `modified_stream` object rather than the hacky `Stream(id=` approach for getting stream objects, where many of the fields wouldn't match the real `Stream` object; just feels safer this way. Backfills are now done in batches of 1k at a time.
- Fixed the `manage.py import_realm` data import tool to ensure it properly imports the new field with a data export. I'm a little disappointed that no tests failed without that fix -- I suspect that this is because `populate_db` doesn't set `creator` for any streams, and thus the end-to-end check that the database ends up in an identical state is not effective. So maybe you can fix that in a quick follow-up PR, as suggested above. I had to do some slight rearranging of the export tool in my own prep commit to make this a 1-line tweak to the existing import code. @mateuszmandera it's worth your taking a quick look to double-check my work on the import tool there.
- Some minor merge conflicts in the CSS. I'm not sure I resolved them entirely correctly, can you please check it carefully and submit a follow-up PR (or fix this PR, if it doesn't merge).

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 6fab5a4e60..65669f670b 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -27,8 +27,7 @@ format used by the Zulip server that they are interacting with.
   [`GET /streams/{stream_id}`](/api/get-stream-by-id),
   [`GET /users/me/subscriptions`](/api/get-subscriptions): Added a new
   field `creator_id`, on stream and subscription objects, containing the
-  user id of the stream creator. This field will be null for streams that
-  do not have a creator.
+  user id of the stream creator.
 
 **Feature level 253**
 
diff --git a/web/styles/input_pill.css b/web/styles/input_pill.css
index 51edf62b9e..4aedabb8d3 100644
--- a/web/styles/input_pill.css
+++ b/web/styles/input_pill.css
@@ -164,7 +164,8 @@
 
 /* These pill are similar to .not-editable, but are meant to show
    profile cards when clicked. */
-.panel_user_list > .display_only_pill,
+.panel_user_list > .pill-container,
+.stream_creator_details > .display_only_pill {
     background-color: hsl(0deg 0% 0% / 7%);
 
     &:hover {
diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
index f484412d11..699282007f 100644
--- a/zerver/lib/import_realm.py
+++ b/zerver/lib/import_realm.py
@@ -970,11 +970,32 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
 
     bulk_import_client(data, Client, "zerver_client")
 
-    # We don't import the Stream and UserGroup models yet, since
-    # they depend on Realm, which isn't imported yet.
-    # But we need the Stream and UserGroup model IDs for
-    # announcements streams and group permissions, respectively
+    # Remap the user IDs for notification_bot and friends to their
+    # appropriate IDs on this server
+    internal_realm = get_realm(settings.SYSTEM_BOT_REALM)
+    crossrealm_user_ids = set()
+    for item in data["zerver_userprofile_crossrealm"]:
+        logging.info(
+            "Adding to ID map: %s %s",
+            item["id"],
+            get_system_bot(item["email"], internal_realm.id).id,
+        )
+        new_user_id = get_system_bot(item["email"], internal_realm.id).id
+        update_id_map(table="user_profile", old_id=item["id"], new_id=new_user_id)
+        crossrealm_user_ids.add(new_user_id)
+        new_recipient_id = Recipient.objects.get(type=Recipient.PERSONAL, type_id=new_user_id).id
+        update_id_map(table="recipient", old_id=item["recipient_id"], new_id=new_recipient_id)
+
+    # We first do a pass of updating model IDs for the cluster of
+    # major models that have foreign keys into each other.
+    # TODO: Should we just do this for all tables at the start?
+    update_model_ids(Realm, data, "realm")
     update_model_ids(Stream, data, "stream")
+    update_model_ids(UserProfile, data, "user_profile")
+    if "zerver_usergroup" in data:
+        update_model_ids(UserGroup, data, "usergroup")
+
+    # Now we prepare to import the Realm table
     re_map_foreign_keys(
         data, "zerver_realm", "new_stream_announcements_stream", related_table="stream"
     )
@@ -983,7 +1004,6 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
         data, "zerver_realm", "zulip_update_announcements_stream", related_table="stream"
     )
     if "zerver_usergroup" in data:
-        update_model_ids(UserGroup, data, "usergroup")
         for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS:
             re_map_foreign_keys(data, "zerver_realm", setting_name, related_table="usergroup")
 
@@ -991,7 +1011,6 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
     # Fix realm subdomain information
     data["zerver_realm"][0]["string_id"] = subdomain
     data["zerver_realm"][0]["name"] = subdomain
-    update_model_ids(Realm, data, "realm")
 
     # Create the realm, but mark it deactivated for now, while we
     # import the supporting data structures, which may take a bit.
@@ -1029,6 +1048,7 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
         # Stream objects are created by Django.
         fix_datetime_fields(data, "zerver_stream")
         re_map_foreign_keys(data, "zerver_stream", "realm", related_table="realm")
+        re_map_foreign_keys(data, "zerver_stream", "creator", related_table="user_profile")
         if role_system_groups_dict is not None:
             # Because the system user groups are missing, we manually set up
             # the defaults for can_remove_subscribers_group for all the
@@ -1046,27 +1066,10 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
         if "zerver_usergroup" not in data:
             set_default_for_realm_permission_group_settings(realm)
 
-    # Remap the user IDs for notification_bot and friends to their
-    # appropriate IDs on this server
-    internal_realm = get_realm(settings.SYSTEM_BOT_REALM)
-    crossrealm_user_ids = set()
-    for item in data["zerver_userprofile_crossrealm"]:
-        logging.info(
-            "Adding to ID map: %s %s",
-            item["id"],
-            get_system_bot(item["email"], internal_realm.id).id,
-        )
-        new_user_id = get_system_bot(item["email"], internal_realm.id).id
-        update_id_map(table="user_profile", old_id=item["id"], new_id=new_user_id)
-        crossrealm_user_ids.add(new_user_id)
-        new_recipient_id = Recipient.objects.get(type=Recipient.PERSONAL, type_id=new_user_id).id
-        update_id_map(table="recipient", old_id=item["recipient_id"], new_id=new_recipient_id)
-
     # To remap foreign key for UserProfile.last_active_message_id
     update_message_foreign_keys(import_dir=import_dir, sort_by_date=sort_by_date)
 
     fix_datetime_fields(data, "zerver_userprofile")
-    update_model_ids(UserProfile, data, "user_profile")
     re_map_foreign_keys(data, "zerver_userprofile", "realm", related_table="realm")
     re_map_foreign_keys(data, "zerver_userprofile", "bot_owner", related_table="user_profile")
     re_map_foreign_keys(
diff --git a/zerver/migrations/0510_stream_creator.py b/zerver/migrations/0511_stream_creator.py
similarity index 66%
rename from zerver/migrations/0510_stream_creator.py
rename to zerver/migrations/0511_stream_creator.py
index c8191af0fa..8f25667034 100644
--- a/zerver/migrations/0510_stream_creator.py
+++ b/zerver/migrations/0511_stream_creator.py
@@ -11,24 +11,27 @@ def backfill_creator_id_from_realm_audit_log(
     apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
 ) -> None:
     RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+    RealmAuditLog.STREAM_CREATED = 601
     Stream = apps.get_model("zerver", "Stream")
 
-    stream_create_logs = RealmAuditLog.objects.filter(
-        event_type=601,  # RealmAuditLog.STREAM_CREATED
-        acting_user_id__isnull=False,
-    ).values_list("modified_stream", "acting_user")
-
     stream_updates = []
-    for stream_id, acting_user_id in stream_create_logs:
-        stream_obj = Stream(id=stream_id, creator_id=acting_user_id)
-        stream_updates.append(stream_obj)
+    for audit_log_entry in RealmAuditLog.objects.select_related("modified_stream").filter(
+        event_type=RealmAuditLog.STREAM_CREATED,
+        acting_user_id__isnull=False,
+    ):
+        assert audit_log_entry.modified_stream is not None
+        stream = audit_log_entry.modified_stream
+        stream.creator_id = audit_log_entry.acting_user_id
+        stream_updates.append(stream)
 
-    Stream.objects.bulk_update(stream_updates, ["creator_id"])
+    Stream.objects.bulk_update(stream_updates, ["creator_id"], batch_size=1000)
 
 
 class Migration(migrations.Migration):
+    atomic = False
+
     dependencies = [
-        ("zerver", "0509_fix_emoji_metadata"),
+        ("zerver", "0510_add_realmauditlog_realm_event_type_index"),
     ]
 
     operations = [
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index ae1b8aaa07..85732c4379 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -19197,12 +19197,12 @@ components:
           description: |
             The ID of the user who created this stream.
 
-            Is `null` for streams with no recorded creator. Streams that existed
-            before this field was introduced and new streams created using
-            backend tools such as `./manage.py create_stream` [management command][management-commands]
-            may not have a creator.
+            Is `null` for streams with no recorded creator, often because they are very
+            old, or were created via a data import tool or [management
+            command][management-commands].
+
+            **Changes**: New in Zulip 9.0 (feature level 254).
 
-            **Changes**: New in Zulip 9.0 (feature level 254)
             [management-commands]: https://zulip.readthedocs.io/en/latest/production/management-commands.html
         invite_only:
           type: boolean
@@ -19936,12 +19936,12 @@ components:
           description: |
             The ID of the user who created this stream.
 
-            Is `null` for streams with no recorded creator. Streams that existed
-            before this field was introduced and new streams created using
-            backend tools such as `./manage.py create_stream` [management command][management-commands]
-            may not have a creator.
+            Is `null` for streams with no recorded creator, often because they are very
+            old, or were created via a data import tool or [management
+            command][management-commands].
+
+            **Changes**: New in Zulip 9.0 (feature level 254).
 
-            **Changes**: New in Zulip 9.0 (feature level 254)
             [management-commands]: https://zulip.readthedocs.io/en/latest/production/management-commands.html
         invite_only:
           type: boolean
```

timabbott commented:
With those changes, this looks great to me, marked to merge once CI passes, thanks @tnmkr!
